python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,302 @@
+    # Observability descriptive message for why the node was launched in the
+    # first place.
+    # TODO
+    launch_reason: Optional[str] = None
+
+    @classmethod
+    def from_node_config(
+        cls,
+        node_config: NodeTypeConfig,
+        status: SchedulingNodeStatus = SchedulingNodeStatus.TO_LAUNCH,
+    ) -> "SchedulingNode":
+        """
+        Create a scheduling node from a node config.
+        """
+        return cls(
+            node_id=cls.next_node_id(),
+            node_type=node_config.name,
+            total_resources=dict(node_config.resources),
+            available_resources=dict(node_config.resources),
+            labels=dict(node_config.labels),
+            status=status,
+        )
+
+    def try_schedule(self, requests: List[ResourceRequest]) -> List[ResourceRequest]:
+        """
+        Try to schedule the resource requests on this node.
+
+        This modifies the node's available resources if the requests are schedulable.
+        When iterating through the requests, the requests are sorted by the
+        `_sort_resource_request` function. The requests are scheduled one by one in
+        the sorted order, and no backtracking is done.
+
+        Args:
+            requests: The resource requests to be scheduled.
+
+        Returns:
+            A list of infeasible requests that cannot be scheduled on this node.
+        """
+        pass
+
+    def compute_score(self) -> UtilizationScore:
+        """
+        Compute the utilization score for this node with respect to the current resource
+        request being scheduled.
+
+        A "higher" score means that this node is more suitable for scheduling the
+        current scheduled resource requests.
+
+        The score is a tuple of 4 values:
+            1. Whether this node is a GPU node and the current resource request has
+                GPU requirements:
+                    0: if this node is a GPU node and the current resource request
+                    placed onto the node has no GPU requirements.
+                    1: if this node is not a GPU node or the current resource request
+                    placed onto the node has GPU requirements.
+            2. The number of resource types being scheduled.
+            3. The minimum utilization rate across all resource types.
+            4. The average utilization rate across all resource types.
+
+        NOTE:
+            This function is adapted from  _resource_based_utilization_scorer from
+            autoscaler v1.
+
+        TODO(rickyx,jjyao):  We should also consider node labels for
+            scoring. For example, if a node has a label that matches the affinity
+            label of the resource request, we should give it a higher score.
+
+        TODO(rickyx): add pluggable scoring functions here.
+
+        Returns:
+            A utilization score for this node.
+        """
+        pass
+
+    @classmethod
+    def next_node_id(cls) -> int:
```

Comment:
hmm not sure what this means

Reply:
just to make sure each `SchduleNode` instance has unique id.